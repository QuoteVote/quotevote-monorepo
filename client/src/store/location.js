/**
 * Location Redux Store
 * Manages user's current location and permission state
 */

import { createSlice } from '@reduxjs/toolkit';
import { getCurrentPosition, checkPermission, GeoLocationError } from '../utils/geolocationService';

const initialState = {
  currentLocation: null, // { lat, lng }
  permissionState: 'prompt', // 'prompt' | 'granted' | 'denied' | 'unsupported'
  isLoading: false,
  error: null, // { type, message }
  lastUpdated: null,
};

const locationSlice = createSlice({
  name: 'location',
  initialState,
  reducers: {
    setLocation: (state, action) => {
      state.currentLocation = action.payload;
      state.lastUpdated = Date.now();
      state.error = null;
    },
    setPermissionState: (state, action) => {
      state.permissionState = action.payload;
    },
    setLoading: (state, action) => {
      state.isLoading = action.payload;
    },
    setError: (state, action) => {
      state.error = action.payload;
      state.isLoading = false;
    },
    clearError: (state) => {
      state.error = null;
    },
    clearLocation: (state) => {
      state.currentLocation = null;
      state.lastUpdated = null;
      state.error = null;
    },
    reset: () => initialState,
  },
});

export const {
  setLocation,
  setPermissionState,
  setLoading,
  setError,
  clearError,
  clearLocation,
  reset,
} = locationSlice.actions;

// Thunks

/**
 * Fetch user's current location
 * @param {object} options - Options passed to getCurrentPosition
 */
export const fetchLocation = (options = {}) => async (dispatch) => {
  dispatch(setLoading(true));
  dispatch(clearError());

  try {
    const location = await getCurrentPosition(options);
    dispatch(setLocation(location));
    dispatch(setPermissionState('granted'));
    return location;
  } catch (error) {
    dispatch(setError({
      type: error.type || GeoLocationError.UNKNOWN,
      message: error.message || 'Failed to get location',
    }));
    
    if (error.type === GeoLocationError.PERMISSION_DENIED) {
      dispatch(setPermissionState('denied'));
    }
    
    throw error;
  } finally {
    dispatch(setLoading(false));
  }
};

/**
 * Check and update permission state
 */
export const updatePermissionState = () => async (dispatch) => {
  try {
    const state = await checkPermission();
    dispatch(setPermissionState(state));
    return state;
  } catch (error) {
    console.error('Failed to check permission:', error);
    return 'prompt';
  }
};

// Selectors

export const selectLocation = (state) => state.location.currentLocation;
export const selectPermissionState = (state) => state.location.permissionState;
export const selectIsLoading = (state) => state.location.isLoading;
export const selectError = (state) => state.location.error;
export const selectLastUpdated = (state) => state.location.lastUpdated;

export const selectHasLocation = (state) => state.location.currentLocation !== null;
export const selectIsPermissionGranted = (state) => state.location.permissionState === 'granted';
export const selectIsPermissionDenied = (state) => state.location.permissionState === 'denied';

export default locationSlice.reducer;
