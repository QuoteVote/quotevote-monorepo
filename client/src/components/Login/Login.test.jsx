import React from 'react'
import { render, act } from '@testing-library/react'
import { vi } from 'vitest'
import { Provider } from 'react-redux'
import store from 'store/store'

// Mock react-hook-form to avoid internal async updates during mount which
// can trigger act() warnings. Use a minimal stub for useForm so tests are
// deterministic.
vi.mock('react-hook-form', () => {
  return {
    useForm: () => ({
      register: () => {},
      handleSubmit: (fn) => (e) => {
        e && e.preventDefault && e.preventDefault()
        return fn()
      },
      errors: {},
      setError: vi.fn(),
      watch: () => true,
    }),
  }
})

// Component being tested
import Login from './Login'

describe('Login test -', () => {
  it('renders correctly', async () => {
    let container
    await act(async () => {
      const res = render(
        <Provider store={store}>
          <Login />
        </Provider>,
      )
      container = res.container
    })
    expect(container.firstChild).toMatchSnapshot()
  })

  it.todo('validate email')
  it.todo('validate password')
  it.todo('submits upon button press')
})
