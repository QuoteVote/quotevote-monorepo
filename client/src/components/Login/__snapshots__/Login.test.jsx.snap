// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login test - > renders correctly 1`] = `
<div
  class="makeStyles-root-1 login-page"
>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-spacing-xs-1 makeStyles-container-2 css-1mydjey-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
    >
      <div
        class="makeStyles-card-12 makeStyles-card-4"
      >
        <div
          class="makeStyles-cardBody-29"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-spacing-xs-2 css-mvfujm-MuiGrid-root"
            justify="space-evenly"
          >
            <div
              class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 makeStyles-header-3 css-1edfpdg-MuiTypography-root"
              >
                Login
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
            >
              <form
                class="makeStyles-formContainer-11"
              >
                <div
                  class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root makeStyles-textfield-7 css-y4n9bg-MuiFormControl-root-MuiTextField-root"
                >
                  <div
                    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedStart css-13p9lry-MuiInputBase-root-MuiOutlinedInput-root"
                  >
                    <div
                      class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-6ga8lj-MuiInputAdornment-root"
                    >
                      <span
                        aria-hidden="true"
                        class="notranslate"
                      >
                        ​
                      </span>
                      <span
                        class="makeStyles-icon-8"
                        data-testid="mui-icon"
                      />
                    </div>
                    <input
                      aria-invalid="false"
                      class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart css-19j5ahi-MuiInputBase-input-MuiOutlinedInput-input"
                      id="username"
                      name="username"
                      placeholder="Email/Username"
                      type="text"
                      value=""
                    />
                    <fieldset
                      aria-hidden="true"
                      class="MuiOutlinedInput-notchedOutline css-1ddm75b-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
                    >
                      <legend
                        class="css-sfzbcg-MuiNotchedOutlined-root"
                      >
                        <span
                          aria-hidden="true"
                          class="notranslate"
                        >
                          ​
                        </span>
                      </legend>
                    </fieldset>
                  </div>
                </div>
                <div
                  class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root makeStyles-textfield-7 css-y4n9bg-MuiFormControl-root-MuiTextField-root"
                >
                  <div
                    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedStart css-13p9lry-MuiInputBase-root-MuiOutlinedInput-root"
                  >
                    <div
                      class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-6ga8lj-MuiInputAdornment-root"
                    >
                      <span
                        aria-hidden="true"
                        class="notranslate"
                      >
                        ​
                      </span>
                      <span
                        class="makeStyles-icon-8"
                        data-testid="mui-icon"
                      />
                    </div>
                    <input
                      aria-invalid="false"
                      class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart css-19j5ahi-MuiInputBase-input-MuiOutlinedInput-input"
                      id="password"
                      name="password"
                      placeholder="Password"
                      type="password"
                      value=""
                    />
                    <fieldset
                      aria-hidden="true"
                      class="MuiOutlinedInput-notchedOutline css-1ddm75b-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
                    >
                      <legend
                        class="css-sfzbcg-MuiNotchedOutlined-root"
                      >
                        <span
                          aria-hidden="true"
                          class="notranslate"
                        >
                          ​
                        </span>
                      </legend>
                    </fieldset>
                  </div>
                </div>
                <label
                  class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-1yyanpp-MuiFormControlLabel-root"
                >
                  <span
                    class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorDefault MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorDefault MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorDefault MuiCheckbox-sizeMedium css-1jviln0-MuiButtonBase-root-MuiSwitchBase-root-MuiCheckbox-root"
                    style="color: gray;"
                  >
                    <input
                      class="PrivateSwitchBase-input css-12xagqm-MuiSwitchBase-root"
                      data-indeterminate="false"
                      name="tos"
                      type="checkbox"
                    />
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-20bmp1-MuiSvgIcon-root"
                      data-testid="CheckBoxOutlineBlankIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                      />
                    </svg>
                  </span>
                  <p
                    class="MuiTypography-root MuiTypography-body2 css-1nmqws6-MuiTypography-root"
                  >
                    I agree to the
                     
                    <a
                      class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways makeStyles-link-9 css-138dq74-MuiTypography-root-MuiLink-root"
                      href="https://github.com/QuoteVote/quotevote-monorepo/blob/main/quote_vote_terms_of_service.md"
                      rel="noopener noreferrer"
                      target="_blank"
                    >
                      Terms of Service
                    </a>
                  </p>
                </label>
                <label
                  class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-1yyanpp-MuiFormControlLabel-root"
                >
                  <span
                    class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorDefault MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorDefault MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorDefault MuiCheckbox-sizeMedium css-1jviln0-MuiButtonBase-root-MuiSwitchBase-root-MuiCheckbox-root"
                    style="color: gray;"
                  >
                    <input
                      class="PrivateSwitchBase-input css-12xagqm-MuiSwitchBase-root"
                      data-indeterminate="false"
                      name="coc"
                      type="checkbox"
                    />
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-20bmp1-MuiSvgIcon-root"
                      data-testid="CheckBoxOutlineBlankIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                      />
                    </svg>
                  </span>
                  <p
                    class="MuiTypography-root MuiTypography-body2 css-1nmqws6-MuiTypography-root"
                  >
                    I agree to the
                     
                    <a
                      class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways makeStyles-link-9 css-138dq74-MuiTypography-root-MuiLink-root"
                      href="https://github.com/QuoteVote/quotevote-monorepo/blob/main/quote_vote_code_of_conduct.md"
                      rel="noopener noreferrer"
                      target="_blank"
                    >
                      Code of Conduct
                    </a>
                  </p>
                </label>
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-colorSecondary MuiButton-fullWidth Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-colorSecondary MuiButton-fullWidth makeStyles-loginButton-5 css-slfb7e-MuiButtonBase-root-MuiButton-root"
                  disabled=""
                  tabindex="-1"
                  type="submit"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 css-1edfpdg-MuiTypography-root"
                  >
                    Log in
                  </p>
                </button>
              </form>
            </div>
            <div
              class="MuiGrid-root MuiGrid-direction-xs-row makeStyles-forgotPasswordContainer-10 css-17fpwt7-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-1edfpdg-MuiTypography-root"
              >
                <a
                  class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways makeStyles-link-9 css-138dq74-MuiTypography-root-MuiLink-root"
                  href="/auth/forgot"
                >
                  Forgot password?
                </a>
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-1edfpdg-MuiTypography-root"
              >
                No account?
                <span
                  style="margin-right: 5px;"
                />
                <a
                  class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways makeStyles-link-9 css-138dq74-MuiTypography-root-MuiLink-root"
                  href="/auth/request-access"
                >
                  Request Access
                </a>
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
